### PR TITLE
Feature/number of requests chart

### LINKED ISSUE #3: Add chart showing number of requests over time
Add a chart showing number of requests over time, similar to the chart on the current APInf dashboard:

screenshot_20170613_121850

Determine whether to use line or bar chart. Choose best aggregation level based on chart timeframe (e.g. daily aggregation for monthly timeframe).

# Resources
The following Elasticsearch aggregation type might be useful for this chart:

- [DateHistogramAggregation](https://www.elastic.co/guide/en/elasticsearch/reference/5.4/search-aggregations-bucket-datehistogram-aggregation.html#search-aggregations-bucket-datehistogram-aggregation)

### PR DESCRIPTION
**Pull request by brylie**
Closes #3

### CODE REVIEW

**Comment by brylie** [comment]
ping @marla-singer

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    .xScale(d3.time.scale())
+    .margin({left: 100, bottom: 100})
+    .showXAxis(true);
+
+  // configure x-axis settings for chart
+  chart.xAxis
+    .axisLabel('Days')
+    .tickFormat(function (d) {
+      // Format tickmarks based on timescale
+      return d3.time.format('%x')(new Date(d))
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: Replace anonymous function to an arrow-function. Make compact function like
```js
.tickFormat( d => d3.time.format('%x')(new Date(d) )
```

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    });
+
+  // configure y-axis settings for chart
+  chart.yAxis
+    .axisLabel('Requests');
+
+  // Fetch Elasticsearch data reactively
+  templateInstance.autorun(function () {
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: Replace to arrow-function

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    const elasticsearchHost = Template.currentData().elasticsearchHost;
+
+    if (elasticsearchHost) {
+      // Get Elasticsearch data
+      Meteor.call('getElasticsearchData', elasticsearchHost, queryParams, function (error, result) {
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: Replace to an arrow-function

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  });
+
+  // Parse chart data reactively
+  templateInstance.autorun(function () {
+    const elasticsearchData = templateInstance.elasticsearchData.get();
+
+    if (elasticsearchData) {
+      // Get aggregations from Elasticsearch data
+      const elasticsearchAggregation = elasticsearchData.aggregations.requests_over_time.buckets;
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: As for me `aggregatedData` name of variable will be more meaning for this case

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      console.log(elasticsearchAggregation);
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: Remove console.log

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  });
+
+  // Render chart reactively
+  templateInstance.autorun(function () {
+    // Get chart data from reactive variable
+    const chartData = templateInstance.chartData.get();
+
+    if (chartData) {
+      console.log(chartData);
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: Remove console.log

File: main/server/methods/elasticsearch.js [hunk 7]
```diff
@@ -79,7 +39,7 @@ Meteor.methods({
             // Throw an error
             throw new Meteor.Error(error.message);
       });
-
+      console.log(elasticsearchData);
```
Comment (marla-singer) [inline main/server/methods/elasticsearch.js]: Remove console.log

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  // Parse chart data reactively
+  templateInstance.autorun(function () {
+    const elasticsearchData = templateInstance.elasticsearchData.get();
+
+    if (elasticsearchData) {
+      // Get aggregations from Elasticsearch data
+      const elasticsearchAggregation = elasticsearchData.aggregations.requests_over_time.buckets;
+      console.log(elasticsearchAggregation);
+      // format object for NVD3 chart
+      const chartDataValues = _.map(elasticsearchAggregation, function (datum) {
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js]: We can escape redundant loop if you will use methods of library properly. 
You must write above:

```js
// Initialize chart
  const chart = nvd3.models.historicalBar();

// configure chart
chart
  .x(d => d.key)
  .y(d => d.doc_count)
  .xScale(d3.time.scale())

...


```

File: dashboard/client/dashboard.html [hunk 9]
```diff
@@ -16,7 +16,15 @@ <h1 class="page-header">API Umbrella Dashboard</h1>
       </div>
 
       <div class="col-sm-8">
-        {{> elasticsearchHealth host=host }}
+        {{> elasticsearchHealth elasticsearchHost=elasticsearchHost }}
+      </div>
+
```
Comment (marla-singer) [inline dashboard/client/dashboard.html]: Remove needless an empty line

File: dashboard/client/requests_over_time/requests_over_time.html [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<template name="requestsOverTime">
+  <h2>Requests over time</h2>
+
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.html:3]: Remove needless an empty line

**Review by marla-singer — CHANGES_REQUESTED** [review]

File: dashboard/client/requests_over_time/requests_over_time.js (hunk 1 above)
Comment (brylie) [inline dashboard/client/requests_over_time/requests_over_time.js]: This code does not need to inherit `this` from the parent context, so an arrow function is not necessary.

File: dashboard/client/requests_over_time/requests_over_time.js (hunk 2 above)
Comment (brylie) [reply dashboard/client/requests_over_time/requests_over_time.js]: This code does not need to inherit this from the parent context, so an arrow function is not necessary.

File: dashboard/client/requests_over_time/requests_over_time.js (hunk 3 above)
Comment (brylie) [reply dashboard/client/requests_over_time/requests_over_time.js]: This code does not need to inherit this from the parent context, so an arrow function is not necessary.

File: dashboard/client/requests_over_time/requests_over_time.html (hunk 10 above)
Comment (brylie) [reply dashboard/client/requests_over_time/requests_over_time.html:3]: In this case, the blank line separates two block elements. Leaving a blank line between blocks improves readability. This is similar to how we leave blank lines between statements in our codebase.

Comment (brylie) [reply dashboard/client/requests_over_time/requests_over_time.html:3]: We can think of the blank line here as a paragraph break.

File: dashboard/client/requests_over_time/requests_over_time.js (hunk 3 above)
Comment (marla-singer) [reply dashboard/client/requests_over_time/requests_over_time.js]: In the main project we have a special rule that says "prefer arrow-function"
As I understand code from this repo will move to main then please follow it and use an arrow-function

Comment (brylie) [reply dashboard/client/requests_over_time/requests_over_time.js]: I did.

File: dashboard/client/requests_over_time/requests_over_time.js [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+  // Render chart reactively
+  templateInstance.autorun(function () {
+    // Get chart data from reactive variable
+    const chartData = templateInstance.chartData.get();
+
+    if (chartData) {
+      // Render the chart with data
+      d3.select('#requests-over-time-chart svg')
+        .datum(chartData)
```
Comment (marla-singer) [inline dashboard/client/requests_over_time/requests_over_time.js:107]: In the future we will need to regulate the size of a chart. To keep it in mind let's create two variables: width and height for canvas chart, and provide them as
```js
d3.select('#requests-over-time-chart svg')
 .attr('width', width)
 .attr('height', height)
  .datum(chartData)
...
```

**Review by marla-singer — APPROVED** [review]

File: dashboard/client/requests_over_time/requests_over_time.js (hunk 11 above)
Comment (brylie) [reply dashboard/client/requests_over_time/requests_over_time.js:107]: Can we work on the chart size in a different PR? I would like to continue working on the wireframes, and we might need to adjust the chart sizes once we have more template structure in place.

**Pull request merged**